File: packages/core/src/symmetry.ts
import type { PitchClass } from './types';
import { ALL_PITCH_CLASSES } from './types';
import { transpose, invert, areEqual } from './pcset';

export function transpositionalStabilizer(pcs: PitchClass[]): PitchClass[] {
  return ALL_PITCH_CLASSES.filter(n => areEqual(transpose(pcs, n), pcs));
}

export function inversionalAxes(pcs: PitchClass[]): PitchClass[] {
  return ALL_PITCH_CLASSES.filter(k => areEqual(invert(pcs, k), pcs));
}

export function stabilizerOrder(pcs: PitchClass[]): number {
  return transpositionalStabilizer(pcs).length + inversionalAxes(pcs).length;
}

export function distinctTranspositions(pcs: PitchClass[]): number {
  return 12 / transpositionalStabilizer(pcs).length;
}

export function abstractGroup(pcs: PitchClass[]): string {
  // Single pitch class (or empty set) has full dihedral symmetry D12
  if (pcs.length <= 1) return 'D12';
  const tOrder = transpositionalStabilizer(pcs).length;
  const iCount = inversionalAxes(pcs).length;
  if (iCount > 0) {
    if (tOrder === 1) return 'Z2';
    return `D${tOrder}`;
  }
  if (tOrder === 1) return 'C1';
  return `C${tOrder}`;
}

File: packages/core-swift/Sources/ChrometriaCore/Symmetry.swift
import Foundation

/// All twelve pitch classes, 0–11. Mirrors `ALL_PITCH_CLASSES` in
/// `packages/core/src/types.ts`.
private let allPitchClasses: [PitchClass] = Array(0..<12)

/// The set of transpositions under which `pcs` is invariant.
public func transpositionalStabilizer(_ pcs: [PitchClass]) -> [PitchClass] {
    allPitchClasses.filter { areEqual(transpose(pcs, $0), pcs) }
}

/// The set of inversion axes under which `pcs` is invariant.
public func inversionalAxes(_ pcs: [PitchClass]) -> [PitchClass] {
    allPitchClasses.filter { areEqual(invert(pcs, $0), pcs) }
}

/// Total order of the symmetry stabilizer (transpositions + inversions).
public func stabilizerOrder(_ pcs: [PitchClass]) -> Int {
    transpositionalStabilizer(pcs).count + inversionalAxes(pcs).count
}

/// Number of distinct transpositions of `pcs` (12 / transpositional stabilizer size).
public func distinctTranspositions(_ pcs: [PitchClass]) -> Int {
    12 / transpositionalStabilizer(pcs).count
}

/// The abstract symmetry group of `pcs`, as a literal group-name string
/// (`"C1"`, `"C2"`, ..., `"D6"`, `"D12"`, `"Z2"`). Mirrors
/// `packages/core/src/symmetry.ts` exactly, including capitalisation —
/// these strings are compared verbatim by later ports.
public func abstractGroup(_ pcs: [PitchClass]) -> String {
    // Single pitch class (or empty set) has full dihedral symmetry D12
    if pcs.count <= 1 { return "D12" }
    let tOrder = transpositionalStabilizer(pcs).count
    let iCount = inversionalAxes(pcs).count
    if iCount > 0 {
        if tOrder == 1 { return "Z2" }
        return "D\(tOrder)"
    }
    if tOrder == 1 { return "C1" }
    return "C\(tOrder)"
}
